documentation/general/torn-sdk.md

```diff
@@ -0,0 +1,142 @@
+| --- | --- | --- |
+| msbuild | From .NET SDK | From .NET SDK |
+| dotnet msbuild | From .NET SDK | From .NET SDK |
+| Visual Studio Design Time | From Visual Studio | From Visual Studio |
+
+Specifically we will be
+
+1. Changing msbuild to use a compiler from the .NET SDK
+2. Changing Visual Studio to use analyzers from Visual Studio
+
+In addition to making our builds more reliable this will also massively simplify our analyzer development strategy. Analyzers following this model can always target the latest Roslyn version without the need for complicated multi-targeting.
+
+## Motivations
+
+There are a number of scenarios where customers end up in a torn state. The most common is when customers have a CI setup with the following items:
+
+1. They use a task like [actions/setup-dotnet][setup-dotnet] to install the .NET SDK and use a flexible version format like `8.x`.
+2. They use `msbuild` to actually drive their build. This `msbuild` comes from the Visual Studio installed on the machine image.
+
+This means that CI systems are updated to the latest .NET SDK virtually as soon as we release them. However the version of Visual Studio is updated much later as CI images usually take several weeks to upgrade to a new Visual Studio version. This is a very common CI setup and means a significant number of our customers end up in a torn state for several weeks.
+
+Another reason is that teams use older Visual Studio versions due to internal constraints: like an organizational policy. At the same time they install the latest .NET SDK which puts them into a torn state.
+
+This also hits any customer that uses a preview version of .NET SDK. These inherently represent a torn SDK state because they almost never match the compiler in Visual Studio This results in blockers for big teams like Bing from testing out our previews.
+
+## Goals
+
+This design has a number of goals:
+
+1. The `msbuild` and `dotnet build` build experience should be equivalent.
+1. The Visual Studio Design time experience is independent of the .NET SDK installed.
+1. To make explicit that it is okay, and even expected, that the design time and command line build experiences can differ when the SDK is in a torn state.
+
+## MSBuild using .NET SDK Compiler
+
+The .NET SDK will start producing a package named Microsoft.NETSdk.Compiler.Roslyn. This will contain a .NET Framework version of the Roslyn compiler that matches .NET Core version. This will be published as a part of the .NET SDK release process meaning it's available for all publicly available .NET SDKs.
+
+The .NET SDK has the capability to [detect a torn state][pr-detect-torn-state]. When this is detected the matching Microsoft.NETSDK.Compiler.Roslyn package for this version of the .NET SDK will be downloaded via `<PackageDownload>`. Then the `$(RoslynTargetsPath)` will be reset to point into the package contents. This will cause MSBuild to load the Roslyn compiler from that location vs. the Visual Studio location.
```

Comment:
do we need to worry about a malicious user overriding RoslynTargetsPath to their own evil version of the compiler?

Reply:
No. Any code which has the ability to override that property today could override the compiler in many other ways. For example you could change `CscToolPath` to change the compiler being used.

Reply:
Another way of thinking about it is that if malicious code has the ability to control MSBuild properties, there are many existing ways that they could hijack the compiler (or other tooling). The change here makes it more centralized but that just makes having a fully functional compiler easier, it doesn't change the attack surface